File: nanobot-security-layer/nanobot_security_layer/pipeline.py
from __future__ import annotations

import logging
from typing import TYPE_CHECKING

from .types import (
    AfterAction,
    AfterResult,
    BeforeAction,
    BeforeResult,
    Finding,
)

if TYPE_CHECKING:
    from .types import ExecutionContext, ExecutionResult, SecurityPlugin

logger = logging.getLogger("nanobot_security_layer")


class SecurityPipeline:

    def __init__(self) -> None:
        self._plugins: list[SecurityPlugin] = []

    def add_plugin(self, plugin: SecurityPlugin) -> None:
        self._plugins.append(plugin)
        self._plugins.sort(key=lambda p: p.priority)

    async def run_before(self, ctx: ExecutionContext) -> BeforeResult:
        all_findings: list[Finding] = []

        for plugin in self._plugins:
            try:
                result = await plugin.before_execute(ctx)
            except Exception:
                logger.exception("Plugin '%s' before_execute error", plugin.id)
                continue

            if result.findings:
                all_findings.extend(result.findings)

            if result.action == BeforeAction.DENY:
                result.findings = all_findings or None
                return result

            if result.action == BeforeAction.MODIFY and result.modified_params:
                ctx.params.update(result.modified_params)

        return BeforeResult(
            action=BeforeAction.ALLOW,
            findings=all_findings or None,
        )

    async def run_after(self, ctx: ExecutionContext, exec_result: ExecutionResult) -> AfterResult:
        final_action = AfterAction.PASS
        final_reason: str | None = None
        final_message: str | None = None
        final_modified: str | None = None
        all_findings: list[Finding] = []

        for plugin in self._plugins:
            try:
                result = await plugin.after_execute(ctx, exec_result)
            except Exception:
                logger.exception("Plugin '%s' after_execute error", plugin.id)
                continue

            if result.findings:
                all_findings.extend(result.findings)

            if result.action == AfterAction.REDACT:
                final_action = AfterAction.REDACT
                final_reason = result.reason
                final_message = result.message
                if result.modified_result is not None:
                    final_modified = result.modified_result
            elif result.action == AfterAction.FLAG and final_action == AfterAction.PASS:
                final_action = AfterAction.FLAG
                final_reason = result.reason
                final_message = result.message

        return AfterResult(
            action=final_action,
            reason=final_reason,
            message=final_message,
            modified_result=final_modified,
            findings=all_findings or None,
        )

    async def destroy(self) -> None:
        for plugin in self._plugins:
            try:
                await plugin.destroy()
            except Exception:
                logger.exception("Plugin '%s' destroy error", plugin.id)
        self._plugins.clear()

File: openclaw-security-layer/src/pipeline.ts
import type {
  SecurityPlugin,
  ExecutionContext,
  ExecutionResult,
  BeforeResult,
  AfterResult,
  Finding,
} from "./types.js";

export class SecurityPipeline {
  private plugins: SecurityPlugin[] = [];

  addPlugin(plugin: SecurityPlugin): void {
    this.plugins.push(plugin);
    this.plugins.sort((a, b) => a.priority - b.priority);
  }

  async runBefore(ctx: ExecutionContext): Promise<BeforeResult> {
    const allFindings: Finding[] = [];

    for (const plugin of this.plugins) {
      if (!plugin.beforeExecute) continue;
      try {
        const result = await plugin.beforeExecute(ctx);
        if (result.findings) allFindings.push(...result.findings);

        if (result.action === "deny") {
          result.findings = allFindings;
          return result;
        }
        if (result.action === "modify" && result.modifiedParams) {
          Object.assign(ctx.params, result.modifiedParams);
        }
      } catch (err) {
        console.error(`[SecurityPipeline] Plugin '${plugin.id}' beforeExecute error:`, err);
      }
    }

    return { action: "allow", findings: allFindings.length ? allFindings : undefined };
  }

  async runAfter(ctx: ExecutionContext, execResult: ExecutionResult): Promise<AfterResult> {
    let finalAction: AfterResult["action"] = "pass";
    let finalMessage: string | undefined;
    let finalModifiedResult: string | undefined;
    let finalReason: string | undefined;
    const allFindings: Finding[] = [];

    for (const plugin of this.plugins) {
      if (!plugin.afterExecute) continue;
      try {
        const result = await plugin.afterExecute(ctx, execResult);
        if (result.findings) allFindings.push(...result.findings);

        if (result.action === "redact") {
          finalAction = "redact";
          finalReason = result.reason;
          finalMessage = result.message;
          if (result.modifiedResult !== undefined) {
            finalModifiedResult = result.modifiedResult;
            execResult = { ...execResult, result: result.modifiedResult };
          }
        } else if (result.action === "flag" && finalAction === "pass") {
          finalAction = "flag";
          finalReason = result.reason;
          finalMessage = result.message;
        }
      } catch (err) {
        console.error(`[SecurityPipeline] Plugin '${plugin.id}' afterExecute error:`, err);
      }
    }

    return {
      action: finalAction,
      reason: finalReason,
      message: finalMessage,
      modifiedResult: finalModifiedResult,
      findings: allFindings.length ? allFindings : undefined,
    };
  }

  async destroy(): Promise<void> {
    for (const plugin of this.plugins) {
      try {
        await plugin.destroy();
      } catch (err) {
        console.error(`[SecurityPipeline] Plugin '${plugin.id}' destroy error:`, err);
      }
    }
    this.plugins = [];
  }
}
